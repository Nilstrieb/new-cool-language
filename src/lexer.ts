import { CompilerError, Span } from "./error";

export type DatalessToken =
  | "kw_function"
  | "kw_let"
  | "p_popen"
  | "p_pclose"
  | "p_bopen"
  | "p_bclose"
  | "p_semi";

export type TokenKind =
  | { kind: DatalessToken }
  | { kind: "identifier"; ident: string }
  | { kind: "lit_string"; value: string };

export type Token = TokenKind & {
  span: Span;
};

export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;

  finish: while (i < input.length) {
    const next = input[i];
    const span: Span = { start: i, end: i + 1 };
    switch (next) {
      case undefined: {
        break finish;
      }
      case "(": {
        tokens.push({ kind: "p_popen", span });
        break;
      }
      case ")": {
        tokens.push({ kind: "p_pclose", span });
        break;
      }
      case "{": {
        tokens.push({ kind: "p_bopen", span });
        break;
      }
      case "}": {
        tokens.push({ kind: "p_bclose", span });
        break;
      }
      case ";": {
        tokens.push({ kind: "p_semi", span });
        break;
      }
      case '"': {
        while (true) {
          const next = input[i + 1];
          span.end++;
          i++;
          if (next === '"') {
            break;
          }
          if (next === undefined) {
            throw new CompilerError(`Unterminated string literal`, span);
          }
        }
        const value = input.slice(span.start + 1, span.end - 1);
        tokens.push({ kind: "lit_string", span, value });
        break;
      }
      default: {
        if (isDigit(next)) {
          throw new Error("digit");
        } else if (isIdentStart(next)) {
          while (isIdentContinue(input[i + 1])) {
            span.end++;
            i++;
          }
          const ident = input.slice(span.start, span.end);
          let kw = isKeyword(ident);
          if (kw) {
            tokens.push({ kind: kw, span });
          } else {
            tokens.push({ kind: "identifier", span, ident: ident });
          }
        } else if (isWhitespace(next)) {
          // ignore
        } else {
          throw new CompilerError(`Invalid character: \`${next}\``, span);
        }
      }
    }

    i++;
  }

  return tokens;
}

function isIdentStart(char: string): boolean {
  return (
    (char <= "Z" && char >= "A") || (char <= "z" && char >= "a") || char === "_"
  );
}

function isIdentContinue(char: string): boolean {
  return (
    (char <= "Z" && char >= "A") ||
    (char <= "z" && char >= "a") ||
    char === "_" ||
    isDigit(char)
  );
}

function isDigit(char: string): boolean {
  return !Number.isNaN(parseInt(char, 10));
}

function isWhitespace(char: string): boolean {
  return char === " " || char === "\t" || char === "\n" || char === "\r";
}

const keywords = new Map<string, DatalessToken>([
  ["function", "kw_function"],
  ["let", "kw_let"],
]);
function isKeyword(kw: string): DatalessToken | undefined {
  return keywords.get(kw);
}
