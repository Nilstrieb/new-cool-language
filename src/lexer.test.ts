import { tokenize } from "./lexer";

it('should tokenize an emtpy function', () => {
    const input = `function hello() {}`;

    const tokens = tokenize(input);

    expect(tokens).toMatchSnapshot();
});

it('should tokenize hello world', () => {
    const input = `print("hello world")`;

    const tokens = tokenize(input);

    expect(tokens).toMatchSnapshot();
});