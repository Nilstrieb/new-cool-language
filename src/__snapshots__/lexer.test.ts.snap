// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should tokenize an emtpy function 1`] = `
[
  {
    "kind": "kw_function",
    "span": {
      "end": 8,
      "start": 0,
    },
  },
  {
    "ident": "hello",
    "kind": "identifier",
    "span": {
      "end": 14,
      "start": 9,
    },
  },
  {
    "kind": "p_popen",
    "span": {
      "end": 15,
      "start": 14,
    },
  },
  {
    "kind": "p_pclose",
    "span": {
      "end": 16,
      "start": 15,
    },
  },
  {
    "kind": "p_bopen",
    "span": {
      "end": 18,
      "start": 17,
    },
  },
  {
    "kind": "p_bclose",
    "span": {
      "end": 19,
      "start": 18,
    },
  },
]
`;

exports[`should tokenize hello world 1`] = `
[
  {
    "ident": "print",
    "kind": "identifier",
    "span": {
      "end": 5,
      "start": 0,
    },
  },
  {
    "kind": "p_popen",
    "span": {
      "end": 6,
      "start": 5,
    },
  },
  {
    "kind": "lit_string",
    "span": {
      "end": 19,
      "start": 6,
    },
    "value": "hello world",
  },
  {
    "kind": "p_pclose",
    "span": {
      "end": 20,
      "start": 19,
    },
  },
]
`;
